### DISCUSSION TITLE
Differences between Quoter and QuoterV2?

### DISCUSSION THREAD

**Discussion by uniyj** [Jeiwan/uniswapv3-book#37] [Milestone 2. Second Swap] [answered]
Just wondering what are the main differences between Uniswap V3's Quoter.sol and QuoterV2.sol? Which quoter contract is the book content [here](https://uniswapv3book.com/docs/milestone_2/quoter-contract/) based on? Thanks!

**Comment by Jeiwan** [comment, marked as answer]
I think I used QuoterV2 as a reference. But my implementation is very minimal. I only wanted to show how trade swaps can be calculated on-chain in Uniswap V3 (and how a mutating function can be turned into a read one).
